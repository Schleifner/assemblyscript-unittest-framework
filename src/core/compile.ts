import { main } from "assemblyscript/asc";
import { join, relative } from "node:path";
import { findRoot } from "../utils/pathResolver.js";

export async function compile(testCodePaths: string[], outputFolder: string, compileFlags: string): Promise<string[]> {
  const pms: Promise<void>[] = [];
  const wasm: string[] = [];
  const root = findRoot(testCodePaths);
  const compile = async (testCodePath: string) => {
    const outputWasm = getNewPath(outputFolder, root, testCodePath).slice(0, -2).concat("wasm");
    wasm.push(outputWasm);
    const outputWat = getNewPath(outputFolder, root, testCodePath).slice(0, -2).concat("wat");
    let ascArgv = [
      testCodePath,
      "--outFile",
      outputWasm,
      "--textFile",
      outputWat,
      "--exportStart",
      "_start",
      "--sourceMap",
      "--debug",
    ];
    if (compileFlags) {
      const argv = compileFlags.split(" ");
      ascArgv = ascArgv.concat(argv);
    }
    const { error, stderr } = await main(ascArgv);
    if (error) {
      // eslint-disable-next-line @typescript-eslint/no-base-to-string
      console.error(stderr.toString());
      throw error;
    }
  };
  for (const testCodePath of testCodePaths) {
    pms.push(compile(testCodePath));
  }
  await Promise.all(pms);
  return wasm;
}

function getNewPath(newFolder: string, oldFolder: string, srcPath: string): string {
  return join(newFolder, relative(oldFolder, srcPath)).replaceAll(/\\/g, "/");
}
