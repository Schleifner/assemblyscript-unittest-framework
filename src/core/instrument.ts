import initInstrumenter from "wasm-instrumentation";
import { InstrumentResult } from "../interface.js";

export async function instrument(sourceWasms: string[], sourceCodePaths: string[]): Promise<InstrumentResult[]> {
  const includeRegexs = sourceCodePaths.map((path) => {
    return `(start:)?${path.slice(0, -3)}.*`;
  });
  const includeFilter = JSON.stringify(includeRegexs);
  const res: InstrumentResult[] = [];
  const instrumenter = await initInstrumenter();
  const pms = sourceWasms.map(async (sourceFile) => {
    const baseName = sourceFile.slice(0, -5);
    const outputFile = baseName.concat(".instrumented.wasm");
    const reportFunction = "covInstrument/traceExpression";
    const sourceMapFile = baseName.concat(".wasm.map");
    const debugInfoFile = baseName.concat(".debugInfo.json");
    const expectInfoFile = baseName.concat(".expectInfo.json");
    const [source, output, report, sourceMap, debugInfo, expectInfo, include] = [
      sourceFile,
      outputFile,
      reportFunction,
      sourceMapFile,
      debugInfoFile,
      expectInfoFile,
      includeFilter,
    ].map((s) => instrumenter.allocateUTF8(s));
    instrumenter._cdc_instrument(source, output, report, sourceMap, expectInfo, debugInfo, include, null, true);
    const result: InstrumentResult = {
      sourceWasm: sourceFile,
      instrumentedWasm: outputFile,
      debugInfo: debugInfoFile,
      expectInfo: expectInfoFile,
    };
    for (const ptr of [source, output, report, sourceMap, debugInfo, expectInfo, include]) {
      instrumenter._free(ptr);
    }
    res.push(result);
  });

  await Promise.all(pms);
  return res;
}
