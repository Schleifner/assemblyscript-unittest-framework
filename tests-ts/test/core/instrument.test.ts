import fs from "fs-extra";
import { join } from "node:path";
import { tmpdir } from "node:os";
import { fileURLToPath, URL } from "node:url";
import { compile } from "../../../src/core/compile.js";
import { instrument } from "../../../src/core/instrument.js";

const fixturePath = join(fileURLToPath(new URL(".", import.meta.url)), "..", "..", "fixture", "constructor.ts");
const outputDir = join(tmpdir(), "assemblyscript-unittest-framework");

test("Instrument", async () => {
  await compile([fixturePath], outputDir, "--memoryBase 16 --exportTable");
  const wasmPath = join(outputDir, "constructor.wasm");
  const sourceCodePath = "tests-ts/fixture/constructor.ts";
  const result = await instrument([wasmPath], [sourceCodePath]);
  expect(result.length).toEqual(1);
  const instrumentedWasm = join(outputDir, "constructor.instrumented.wasm");
  const debugInfo = join(outputDir, "constructor.debugInfo.json");
  const expectInfo = join(outputDir, "constructor.expectInfo.json");
  expect(result[0]).toEqual({
    sourceWasm: wasmPath,
    instrumentedWasm,
    debugInfo,
    expectInfo,
  });
  expect(fs.existsSync(instrumentedWasm)).toEqual(true);
  expect(fs.existsSync(debugInfo)).toEqual(true);
  expect(fs.existsSync(expectInfo)).toEqual(true);
  expect(fs.readFileSync(debugInfo, { encoding: "utf8" })).toMatchSnapshot();
});
