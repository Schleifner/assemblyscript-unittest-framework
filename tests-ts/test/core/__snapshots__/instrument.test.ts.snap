// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Instrument 1`] = `"{"debugFiles":["~lib/rt/common.ts","~lib/rt/tlsf.ts","~lib/shared/typeinfo.ts","~lib/rt/itcms.ts","tests-ts/fixture/constructor.ts","~lib/shared/runtime.ts","~lib/util/number.ts","~lib/util/math.ts","~lib/util/string.ts","~lib/rt.ts","~lib/util/error.ts"],"debugInfos":{"tests-ts/fixture/constructor/Foo#constructor":{"branchInfo":[[0,1],[0,2]],"index":0,"lineInfo":[[],[],[[4,7,25],[4,8,24],[4,11,4],[4,11,16]]]}}}"`;
